File: 528-c/528.按权重随机选择.c
/*
 * @lc app=leetcode.cn id=528 lang=c
 *
 * [528] 按权重随机选择
 */

// @lc code=start
// #include <stdlib.h>

typedef struct {
    int *sum;
    int len;
} Solution;


Solution* solutionCreate(int* w, int wSize) {
    Solution *obj = malloc(sizeof(Solution));
    obj->sum = malloc(sizeof(int)*wSize);
    obj->len = wSize;
    
    obj->sum[0] = w[0];
    for (int i = 1; i < wSize; i++) {
        obj->sum[i] = obj->sum[i-1] + w[i];
    }

    return obj;
}

// 返回target是在obj->sum的第几个区间
// 寻找第一个sum[i] >= target的下标i
int binSearch(Solution *obj, int target) {
    int l = 0,r = obj->len-1;
    for (; l < r; ) {
        int mid = (r - l) / 2 + l;
        if (target <= obj->sum[mid]) {
            r = mid;
        }else {
            l = mid+1;
        }
    }
    return l;
}
int solutionPickIndex(Solution* obj) {
    int max = obj->sum[obj->len-1];
    // 得到的数据范围是[1,max]
    int num = rand() % max + 1;
    return binSearch(obj,num);
}

void solutionFree(Solution* obj) {
    free(obj->sum);
    free(obj);
}

// int main() {
//     int w[1] = {1};
//     Solution *obj = solutionCreate(w,1);
//     printf("%d\n", solutionPickIndex(obj));
// }
/**
 * Your Solution struct will be instantiated and called as such:
 * Solution* obj = solutionCreate(w, wSize);
 * int param_1 = solutionPickIndex(obj);
 
 * solutionFree(obj);
*/
// @lc code=end

File: 528.按权重随机选择.go
/*
 * @lc app=leetcode.cn id=528 lang=golang
 *
 * [528] 按权重随机选择
 */

// @lc code=start
package main

import (
	"math/rand"
	"sort"
)

// 这里有一个问题，w[i] <= 100000,w.length <= 10000,
// 因此w数据总和最大10的9次方，可能会超过int范围
// 并且使用的随机函数rand.Intn也只返回int类型，可能无法覆盖sum的最大值，但是最后还是通过了
type Solution struct {
	sum []int
}

func Constructor(w []int) Solution {
	n := len(w)
	sum := make([]int,n+1)
	sum[0] = 0
	for i := 1 ; i <= n ; i++ {
		sum[i] = sum[i-1] + w[i-1]
	}
	return Solution{sum}
}

func (this *Solution) PickIndex() int {
	n := len(this.sum)
	max := this.sum[n-1]
	num := rand.Intn(max)
	return sort.Search(n,func(i int) bool {return this.sum[i] > num})-1
}


/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(w);
 * param_1 := obj.PickIndex();
 */
// @lc code=end

